server/src/graql/gremlin/fragment/AbstractRolePlayerFragment.java

```diff
@@ -83,9 +88,34 @@ final String innerName() {
     }
 
     @Override
-    public final Set<Weighted<DirectedEdge<Node>>> directedEdges(
-            Map<NodeId, Node> nodes, Map<Node, Map<Node, Fragment>> edges) {
-        return directedEdges(edge(), nodes, edges);
+    public Set<Node> getNodes() {
+        Node start = new Node(NodeId.of(NodeId.NodeType.VAR, start()));
+        Node end = new Node(NodeId.of(NodeId.NodeType.VAR, end()));
+        Node middle = new Node(NodeId.of(NodeId.NodeType.VAR, edge()));
+        middle.setInvalidStartingPoint();
+        return new HashSet<>(Arrays.asList(start, end, middle));
+    }
+
+    @Override
+    public final Set<Weighted<DirectedEdge>> directedEdges(Map<NodeId, Node> nodes,
+                                                           Map<Node, Map<Node, Fragment>> edges) {
+
+        // this is a somewhat special case, where the middle node being converted to a vertex
+        // may be addressed by a variable
+
+        Node start = nodes.get(NodeId.of(NodeId.NodeType.VAR, start()));
+        Node end = nodes.get(NodeId.of(NodeId.NodeType.VAR, end()));
+        Node middle = nodes.get(NodeId.of(NodeId.NodeType.VAR, edge()));
+        middle.setInvalidStartingPoint();
+
+        if (!edges.containsKey(middle)) {
```

Comment:
just wondering, what's the scenario when edges contain `middle` key?

Reply:
This is a remaining piece of procedural code that modifies arguments in place - `edges` actually maps from `middle` node to `Fragment`s and the mapping is created here. Hoping to remove.